dask/bag/core.py

```diff
@@ -2498,3 +2478,121 @@ def to_dataframe(seq, columns, dtypes):
         seq = list(seq)
     res = pd.DataFrame(seq, columns=list(columns))
     return res.astype(dtypes, copy=False)
+
+
+def repartition_npartitions(bag, npartitions):
+    """ Changes the number of partitions of the bag.
+
+    This can be used to reduce or increase the number of partitions
+    of the bag.
+    """
+    new_name = "repartition-%d-%s" % (npartitions, tokenize(bag, npartitions))
+    if npartitions == bag.npartitions:
+        return bag
+    elif bag.npartitions > npartitions:
+        ratio = bag.npartitions / npartitions
+        new_partitions_boundaries = [
+            int(old_partition_index * ratio)
+            for old_partition_index in range(npartitions + 1)
+        ]
+        return _repartition_from_boundaries(bag, new_partitions_boundaries, new_name)
+    else:  # npartitions > bag.npartitions
+        div, mod = divmod(npartitions, bag.npartitions)
+        nsplits = [div] * bag.npartitions
+        nsplits[-1] += mod
+        return _split_partitions(bag, nsplits, new_name)
+
+
+def repartition_size(bag, size):
+    """
+    Repartition bag so that new partitions have approximately `size` memory usage each
+    """
+    if isinstance(size, str):
+        size = parse_bytes(size)
+    size = int(size)
+    mem_usages = bag.map_partitions(total_mem_usage).compute()
+
+    # 1. split each partition that is larger than partition size
+    nsplits = [1 + mem_usage // size for mem_usage in mem_usages]
+    if any((nsplit > 1 for nsplit in nsplits)):
+        split_name = "repartition-split-{}-{}".format(size, tokenize(bag))
+        bag = _split_partitions(bag, nsplits, split_name)
+        # update mem_usages to account for the split partitions
+        split_mem_usages = []
+        for n, usage in zip(nsplits, mem_usages):
+            split_mem_usages.extend([usage / n] * n)
+        mem_usages = split_mem_usages
+
+    # 2. now that all partitions are less than size, concat them up to size
+    assert all((mem_usage <= size for mem_usage in mem_usages))
+    new_npartitions = list(map(len, iter_chunks(mem_usages, size)))
+    new_partitions_boundaries = accumulate(operator.add, new_npartitions)
+    new_name = "repartition-{}-{}".format(size, tokenize(bag))
+    return _repartition_from_boundaries(bag, new_partitions_boundaries, new_name)
+
+
+def total_mem_usage(bag):
+    return sys.getsizeof(bag)
```

Comment:
You might want to try `dask.sizeof.sizeof` instead here.  I don't think that `sys.getsizeof` has the behavior that you want.

```python
In [1]: import sys                                                                                 

In [2]: sys.getsizeof(list(range(1000000)))                                                        
Out[2]: 9000120

In [3]: sys.getsizeof({"x": list(range(1000000))})                                                 
Out[3]: 248
```

```python
In [6]: from dask.sizeof import sizeof                                                             

In [7]: sizeof({"x": list(range(1000000))})                                                        
Out[7]: 37000482
```